thrift/src/main/java/com/linecorp/armeria/server/thrift/ThriftCallService.java

```diff
@@ -123,28 +124,29 @@ public RpcResponse serve(ServiceRequestContext ctx, RpcRequest call) throws Exce
 
         // Ensure that such a service exists.
         final ThriftServiceEntry e = entries.get(serviceName);
+        final RpcResponse reply = JavaVersionSpecific.get().newRequestContextAwareRpcResponse(ctx);
         if (e != null) {
             // Ensure that such a method exists.
             final ThriftFunction f = e.metadata.function(method);
             if (f != null) {
                 if (f.implementation() != null) {
-                    final DefaultRpcResponse reply = new DefaultRpcResponse();
-                    invoke(ctx, f.implementation(), f, call.params(), reply);
+                    invoke(ctx, f.implementation(), f, call.params(), reply.toCompletableFuture());
                     return reply;
                 }
                 // Should never reach here because of the way ThriftServiceEntry is created
-                return new DefaultRpcResponse(new TApplicationException(
+                reply.completionFuture().completeExceptionally(new TApplicationException(
                         TApplicationException.UNKNOWN, "null implementation: " + call.method()));
+                return reply;
             }
         }
-
-        return new DefaultRpcResponse(new TApplicationException(
+        reply.toCompletableFuture().completeExceptionally(new TApplicationException(
                 TApplicationException.UNKNOWN_METHOD, "unknown method: " + call.method()));
+        return reply;
     }
 
     private static void invoke(
             ServiceRequestContext ctx,
-            Object impl, ThriftFunction func, List<Object> args, DefaultRpcResponse reply) {
+            Object impl, ThriftFunction func, List<Object> args, CompletableFuture<Object> reply) {
```

Comment:
I wonder that there are any reasons to define the type parameter of `CompletableFuture` as a `Object`?

or how about using `CompletableFuture<?>`?

Reply:
I cannot complete the future here
```java
public void onComplete(Object response) {
    reply.complete(response);
}
```
So let me just use `Object`. 😉